.github/scripts/toolchain/group-gcc-test-failures.py

```diff
@@ -0,0 +1,149 @@
+        reading_init_lines = True
+
+        with open(log_file, 'r', encoding="utf8") as file:            
+            log_paragraph = []
+            for line in file:
+                # Skip the initial lines that are not part of output of any unittest
+                if reading_init_lines:
+                    if line.startswith("Running ") and line.rstrip().endswith(".exp ..."):
+                        reading_init_lines = False                        
+                    continue
+
+                log_paragraph.append(line)
+
+                log_type_candidate = line.split(":")[0]
+                if log_type_candidate in FAIL_LOG_TYPES:
+                    # Store log messages from tests that failed
+                    log_messages[log_type_candidate].append(log_paragraph)
+                    log_paragraph = []
+                elif log_type_candidate in PASS_LOG_TYPES:
+                    # Ignore log messages from tests that passed
+                    log_paragraph = []
+
+    print("Finished reading all log files.")
+    print()
+    print("Number of found tests for each result type:")
+    for log_type, messages in log_messages.items():
+        print(f"  {log_type: <12} {len(messages)}")
+    print()
+
+    return log_messages
+
+
+def print_path_structure(log_messages: LOG_MESSAGES_TYPE, summary_output: str) -> None:
+    def parse_filepath(message: list[str]) -> list[str]:
+        # We parse here the last line of each log message, example:
+        # FAIL: g++.dg/contracts/contracts-pre10.C   20 blank line(s) in output
+        assert any(message[-1].startswith(log_type) for log_type in FAIL_LOG_TYPES)
+        filepath = message[-1].split(" ")[1]
+        # The path might but don't have to start with tool name split by dot
+        if "." in filepath and "/" not in filepath.split(".", 1)[0]:
+            tool_name, filepath = filepath.split(".", 1)
+            return [tool_name] + filepath.split("/")
+        return filepath.split("/")
+    
+    @dataclasses.dataclass
+    class PathSegmentStats:
+        count: int = 0
+        children: dict[str, "PathSegmentStats"] = dataclasses.field(default_factory=dict)
+
+    def print_path_segment_stats(
+        entry: PathSegmentStats,
+        print_limit: int = 0,
+        print_markdown_list: bool = False,
+        indent: int = 0
+    ) -> None:
+        """Prints hierarchical path segments with their occurrence counts
+        
+        print_limit: Print only path segments with occurrence count >= print_limit
+        """
+        sorted_path_stats = sorted(entry.children.items(), key=lambda x: x[1].count, reverse=True)
+        for key, value in sorted_path_stats:
+            if value.count < print_limit:
+                if not print_markdown_list:
+                    print(" " * (indent) + "...")
+                return
+            
+            if print_markdown_list:
+                print(" " * indent + f"- {key}: {value.count}")
```

Comment:
Just an idea: Is there a tool or format, e.g. Markdown, that would alllow seeing the output file with ability to collapse/expand the items?

Reply:
Could be done in markdown https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab#33-nested-collapsible-sections which is probably easiest to view. Do you suggest to write whole output to `GITHUB_STEP_SUMMARY` just with collapsed sections?

Reply:
I don't have any preference regarding that, downloading the artifcact and viewing it, e.g. in VSCode if fine for me. There might be limit for the `GITHUB_STEP_SUMMARY` size.

Reply:
Added collapsible sections to the output.

Example of output from debugging data: [output.md](https://github.com/user-attachments/files/16784158/output.md)

Example from CI will be available here: https://github.com/Windows-on-ARM-Experiments/mingw-woarm64-build/actions/runs/10597824698

Reply:
Very nice!